torch/csrc/jit/tensorexpr/registerizer.h

```diff
@@ -0,0 +1,448 @@
+    if (v->buf() != info_->buf) {
+      return IRMutator::mutate(v);
+    }
+
+    if (initializerReady_ && info_->parent == v->get_parent()) {
+      initializer_ = v;
+      initializerReady_ = false;
+      // This is the easiest way to return an empty statement;
+      return new Block({});
+    }
+
+    initializerReady_ = false;
+
+    // sanity check indices for the same buf must have the same dimensionality.
+    assert(v->indices().size() == info_->indices.size());
+    for (int i = 0; i < info_->indices.size(); ++i) {
+      if (!exprEquals(v->indices()[i], info_->indices[i])) {
+        return IRMutator::mutate(v);
+      }
+    }
+    const Expr* new_val = v->value()->accept_mutator(this);
+
+    Store* s = new Store(var_wrapper_, {}, new_val, v->mask());
+    return s;
+  }
+
+  // Finds the Stmt in parent which contains stmt.
+  const Stmt* findInsertionPoint(const Stmt* stmt, const Block* parent) {
+    while (stmt) {
+      if (stmt->get_parent() == parent) {
+        return stmt;
+      }
+      stmt = stmt->get_parent();
+    }
+    return nullptr;
+  }
+
+  Stmt* mutate(const Block* v) override {
+    // We need to mutate this block in place, rather than clone - since other
+    // AccessInfo objects may hold a pointer to it.
+    Block* v1 = const_cast<Block*>(v); // NOLINT
+    assert(v1);
+
+    Stmt* first_changed{nullptr};
+    Stmt* last_changed{nullptr};
+    std::list<Stmt*> stmts = v1->stmts();
+    for (Stmt* stmt : stmts) {
+      dirty_ = false;
+      Stmt* stmt_new = stmt->accept_mutator(this);
+      if (dirty_) {
+        first_changed = first_changed ? first_changed : stmt_new;
+        last_changed = stmt_new;
+      }
+
+      if (stmt_new == stmt) {
+        continue;
+      }
+      v1->replace_stmt(stmt, stmt_new);
+      first_changed = first_changed ? first_changed : stmt_new;
+      last_changed = stmt_new;
+    }
+
+    dirty_ = first_changed != nullptr;
+
+    if (v != info_->parent) {
+      return v1;
+    }
+
+    Stmt* let;
+    // If we didn't find an initial store: intialize with the original buffer.
+    if (!initializer_) {
+      let = new Let(
+          var_,
+          new Load(
+              info_->buf->base_handle()->dtype(),
+              info_->buf,
+              info_->indices,
+              new IntImm(1)));
+    } else {
+      let = new Let(var_, initializer_->value());
+    }
+    v1->insert_stmt_before(let, first_changed);
+
+    // If it was written to the buffer, make sure we write it out.
+    if (info_->stores.size() > 0) {
+      v1->insert_stmt_after(
+          new Store(info_->buf, info_->indices, var_, new IntImm(1)),
+          last_changed);
+    }
+    return v1;
+  }
+
+ private:
+  std::shared_ptr<AccessInfo> info_;
+  Var* var_;
+  Buf* var_wrapper_;
+  const Store* initializer_;
+  bool dirty_{false};
+  bool initializerReady_{true};
+};
+
+// Apply scalar replacement to all accesses in s.
+TORCH_API Stmt* registerize(Stmt* s) {
```

Comment:
Minor: since we are in the stage of moving around different passes and try different orders. It will be good to list some of the ordering requirement with CudaCodeGen here. For example: this must be invoked after threadIdx flattening, but must happen before pass xyz. It doesn't have to be complete, just enough to remind us where not to move this into.